File: lib/core/file_dialog.ml
open Core

let validate_label value =
  if
    String.length value > 4096
    || String.is_empty (String.strip value)
    || String.contains value '\000'
    || not (Stdlib.String.is_valid_utf_8 value)
  then
    Or_error.error_string "file dialog labels must be bounded nonblank UTF-8 without NUL"
  else Ok ()
;;

module Open = struct
  module Selection = struct
    type t =
      | Files
      | Directories
      | Files_and_directories
    [@@deriving equal, sexp_of]
  end

  type t =
    { selection : Selection.t
    ; multiple : bool
    ; title : string
    ; accept_label : string
    ; directory : File_path.t option
    }
  [@@deriving equal, sexp_of]

  let create
        ?(selection = Selection.Files)
        ?(multiple = false)
        ?(title = "Open")
        ?(accept_label = "Open")
        ?directory
        ()
    =
    let open Or_error.Let_syntax in
    let%bind () = validate_label title in
    let%map () = validate_label accept_label in
    { selection; multiple; title; accept_label; directory }
  ;;

  let selection t = t.selection
  let multiple t = t.multiple
  let title t = t.title
  let accept_label t = t.accept_label
  let directory t = t.directory
end

module Save = struct
  type t =
    { directory : File_path.t
    ; suggested_name : string
    ; title : string
    ; accept_label : string
    }
  [@@deriving equal, sexp_of]

  let create ~directory ~suggested_name ?(title = "Save") ?(accept_label = "Save") () =
    let open Or_error.Let_syntax in
    let%bind () = validate_label title in
    let%bind () = validate_label accept_label in
    if
      String.is_empty suggested_name
      || String.length suggested_name > 255
      || String.contains suggested_name '/'
      || String.contains suggested_name '\000'
      || String.equal suggested_name "."
      || String.equal suggested_name ".."
      || not (Stdlib.String.is_valid_utf_8 suggested_name)
    then
      Or_error.error_string
        "suggested name must be one UTF-8 filename of 1..255 bytes without slash or NUL"
    else Ok { directory; suggested_name; title; accept_label }
  ;;

  let directory t = t.directory
  let suggested_name t = t.suggested_name
  let title t = t.title
  let accept_label t = t.accept_label
end

module Error = Gpuio_protocol.Wire.File_dialog.Error

module Request = struct
  type t =
    | Open of Open.t
    | Save of Save.t
  [@@deriving equal, sexp_of]
end

module Expert = struct
  module Wire = Gpuio_protocol.Wire.File_dialog

  let to_wire = function
    | Request.Open config ->
      let selection : Wire.Selection.t =
        match Open.selection config with
        | Files -> Files
        | Directories -> Directories
        | Files_and_directories -> Files_and_directories
      in
      Wire.Config.Open
        { selection
        ; multiple = Open.multiple config
        ; title = Open.title config
        ; accept_label = Open.accept_label config
        ; directory = Option.map (Open.directory config) ~f:File_path.to_string
        }
    | Request.Save config ->
      Wire.Config.Save
        { directory = File_path.to_string (Save.directory config)
        ; suggested_name = Save.suggested_name config
        ; title = Save.title config
        ; accept_label = Save.accept_label config
        }
  ;;

  let result_of_wire request = function
    | Wire.Result.Failed error -> Error error
    | Cancelled -> Ok None
    | Selected paths ->
      let maximum =
        match request with
        | Request.Open config -> if Open.multiple config then 128 else 1
        | Save _ -> 1
      in
      if
        List.is_empty paths
        || List.length paths > maximum
        || List.sum (module Int) paths ~f:String.length > 262_144
      then Error Error.Native_failure
      else
        List.map paths ~f:File_path.of_string
        |> Or_error.combine_errors
        |> Result.map ~f:Option.some
        |> Result.map_error ~f:(fun _ -> Error.Native_failure)
  ;;
end

File: rust/protocol/src/file_dialog.rs
//! Validated data for native open/save requests. Request routing and native panel
//! ownership belong to the runtime, not this module.
use crate::file_path::FilePath;
use binprot::macros::BinProtWrite;

pub const MAX_SELECTED_PATHS: usize = 128;
pub const MAX_SELECTED_PATH_BYTES: usize = 262_144;

#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub enum FileSelection {
    Files,
    Directories,
    FilesAndDirectories,
}

#[derive(Clone, Debug, PartialEq, Eq, BinProtWrite)]
pub struct OpenFileConfig {
    pub selection: FileSelection,
    pub multiple: bool,
    pub title: String,
    pub accept_label: String,
    pub directory: Option<FilePath>,
}

#[derive(Clone, Debug, PartialEq, Eq, BinProtWrite)]
pub struct SaveFileConfig {
    pub directory: FilePath,
    pub suggested_name: String,
    pub title: String,
    pub accept_label: String,
}

#[derive(Clone, Debug, PartialEq, Eq, BinProtWrite)]
pub enum FileDialogConfig {
    Open(OpenFileConfig),
    Save(SaveFileConfig),
    Capabilities,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub enum FileSelectionSupport {
    Unsupported,
    Single,
    Multiple,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub struct FileDialogCapabilities {
    pub files: FileSelectionSupport,
    pub directories: FileSelectionSupport,
    pub files_and_directories: FileSelectionSupport,
    pub save: bool,
}

fn label(value: &str) -> bool {
    crate::v1::CommandConfig::valid_text(value, 4096)
}

impl FileDialogConfig {
    pub fn is_valid(&self) -> bool {
        match self {
            Self::Capabilities => true,
            Self::Open(config) => label(&config.title) && label(&config.accept_label),
            Self::Save(config) => {
                label(&config.title)
                    && label(&config.accept_label)
                    && !config.suggested_name.is_empty()
                    && config.suggested_name.len() <= 255
                    && !config.suggested_name.contains(['/', '\0'])
                    && config.suggested_name != "."
                    && config.suggested_name != ".."
            }
        }
    }

    /// Validate the entire selection without silently truncating or dropping a
    /// path. FilePath already enforces absolute, bounded, NUL-free native bytes.
    pub fn accepts_selection(&self, paths: &[FilePath]) -> bool {
        let maximum = match self {
            Self::Capabilities => return false,
            Self::Open(config) if config.multiple => MAX_SELECTED_PATHS,
            Self::Open(_) | Self::Save(_) => 1,
        };
        !paths.is_empty()
            && paths.len() <= maximum
            && paths
                .iter()
                .map(|path| path.as_bytes().len())
                .sum::<usize>()
                <= MAX_SELECTED_PATH_BYTES
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub enum FileDialogError {
    InvalidRequest,
    Unsupported,
    Busy,
    Closed,
    NotReady,
    NativeFailure,
    LimitExceeded,
}

#[derive(Clone, Debug, PartialEq, Eq, BinProtWrite)]
pub enum FileDialogResult {
    Selected(Vec<FilePath>),
    Cancelled,
    Failed(FileDialogError),
    Capabilities(FileDialogCapabilities),
}

#[cfg(test)]
mod tests {
    use super::*;

    fn open(multiple: bool) -> FileDialogConfig {
        FileDialogConfig::Open(OpenFileConfig {
            selection: FileSelection::Files,
            multiple,
            title: "Open".into(),
            accept_label: "Choose".into(),
            directory: None,
        })
    }

    #[test]
    fn selection_limits_reject_the_whole_result_instead_of_truncating() {
        let path = FilePath::new(b"/tmp/\xff.txt".to_vec()).unwrap();
        assert!(!open(true).accepts_selection(&[]));
        assert!(open(false).accepts_selection(std::slice::from_ref(&path)));
        assert!(!open(false).accepts_selection(&[path.clone(), path.clone()]));
        assert!(open(true).accepts_selection(&vec![path.clone(); MAX_SELECTED_PATHS]));
        assert!(!open(true).accepts_selection(&vec![path; MAX_SELECTED_PATHS + 1]));
        let path = FilePath::new(vec![b'/'; crate::file_path::MAX_PATH_BYTES]).unwrap();
        assert!(open(true).accepts_selection(&vec![path.clone(); 16]));
        assert!(!open(true).accepts_selection(&vec![path; 17]));
    }

    #[test]
    fn labels_and_filename_hints_are_validated_without_rewriting() {
        let directory = FilePath::new(b"/tmp".to_vec()).unwrap();
        for name in ["a.sql.s", ".hidden", "λ.txt", " "] {
            let config = FileDialogConfig::Save(SaveFileConfig {
                directory: directory.clone(),
                suggested_name: name.into(),
                title: "Save".into(),
                accept_label: "Choose".into(),
            });
            assert!(config.is_valid());
            assert!(!config.accepts_selection(&[directory.clone(), directory.clone()]));
        }
        for name in ["", ".", "..", "a/b", "a\0b", &"a".repeat(256)] {
            let config = FileDialogConfig::Save(SaveFileConfig {
                directory: directory.clone(),
                suggested_name: name.into(),
                title: "Save".into(),
                accept_label: "Choose".into(),
            });
            assert!(!config.is_valid());
        }
        for title in ["", " ", "\0", &"a".repeat(4097)] {
            let FileDialogConfig::Open(mut config) = open(false) else {
                unreachable!()
            };
            config.title = title.into();
            assert!(!FileDialogConfig::Open(config).is_valid());
        }
    }
}
